### PR TITLE
feat(daemon): allows daemonization options to be fetched from app settings

### PR DESCRIPTION
**Pull request by noirbizarre**
## Description

This pull-requests allows to provide daemonization settings (`--logfile`, `--pidfile`, `--uid`, `--gid`, `--umask` and `--executable`) from your config file with the following new settings:
- `beat_logfile`
- `beat_pidfile`
- `beat_uid`
- `beat_gid`
- `beat_umask`
- `beat_executable`
- `worker_logfile`
- `worker_pidfile`
- `worker_uid`
- `worker_gid`
- `worker_umask`
- `worker_executable`
- `events_logfile`
- `events_pidfile`
- `events_uid`
- `events_gid`
- `events_umask`
- `events_executable`

This PR is backward compatible, tested and documented. Missing CLI documentation for daemonization settings have been added in the process.

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
thanks for the PR. can you check the CI failures please?

**Comment by noirbizarre** [comment]
There it is, PR fixed, Ci is all green 👍🏼 
(last `Celery / Unit 3.9` job is stuck but I'be dealing with this issue for hours: github actions cache retrieval is not in a good state, it should unlock any time)

**Comment by auvipy** [comment]
What extra benefits the changes will provide?

**Comment by noirbizarre** [comment]
Today, the only way to provide those values are from cli flag which prevent retrieving those values from any registry (unless you have a way to fetch those values before running those daemons and passing the processed result to the cli, which is not possible in my case).
This change allows to specify those values from settings (like most of the settings you can set on cli), meaning from Python, allowing for dynamic values (which is my use case) and any kind of preprocessing.
This also works for static settings in the case you have a templated celeryconf.py loaded with `Celery.config_from_object()` (the exact case which is tested in this PR and was not possible without this PR)

**Comment by auvipy** [comment]
yeah that looks nice. I am in favor of it. Lets us wait for other maintainers to look into it. I would appreciate if in the mean time you could review this PR https://github.com/celery/celery/pull/8489

File: docs/userguide/configuration.rst [hunk 1]
```diff
@@ -3180,6 +3180,61 @@ Message serialization format used when sending event messages.
...
+An optionnal file path for :program:`celery events` to log into (default to `stdout`).
+
+.. setting:: events_pidfile
+
+``events_pidfile``
+~~~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal file path for :program:`celery events` to create/store it PID file (default to no PID file created).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3199]:
```suggestion
An optional file path for :program:`celery events` to create/store its PID file (default to no PID file created).
```

File: docs/userguide/configuration.rst [hunk 2]
```diff
@@ -3180,6 +3180,61 @@ Message serialization format used when sending event messages.
...
 
 
+.. setting:: events_logfile
+
+``events_logfile``
+~~~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal file path for :program:`celery events` to log into (default to `stdout`).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3190]:
```suggestion
An optional file path for :program:`celery events` to log into (defaults to `stdout`).
```

File: docs/userguide/configuration.rst [hunk 3]
```diff
@@ -3180,6 +3180,61 @@ Message serialization format used when sending event messages.
...
+An optionnal file path for :program:`celery events` to create/store it PID file (default to no PID file created).
+
+.. setting:: events_uid
+
+``events_uid``
+~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal user ID to use when events :program:`celery events` drops its privileges (default to no UID change).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3208]:
```suggestion
An optional user ID to use when events :program:`celery events` drops its privileges (defaults to no UID change).
```

File: docs/userguide/configuration.rst [hunk 4]
```diff
@@ -3180,6 +3180,61 @@ Message serialization format used when sending event messages.
...
+
+.. setting:: events_gid
+
+``events_gid``
+~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal group ID to use when :program:`celery events` daemon drops its privileges (default to no GID change).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3217]:
```suggestion
An optional group ID to use when :program:`celery events` daemon drops its privileges (defaults to no GID change).
```

File: docs/userguide/configuration.rst [hunk 5]
```diff
@@ -3180,6 +3180,61 @@ Message serialization format used when sending event messages.
...
+
+.. setting:: events_umask
+
+``events_umask``
+~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal `umask` to use when :program:`celery events` creates files (log, pid...) when daemonizing.
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3226]:
```suggestion
An optional `umask` to use when :program:`celery events` creates files (log, pid...) when daemonizing.
```

File: docs/userguide/configuration.rst [hunk 6]
```diff
@@ -3180,6 +3180,61 @@ Message serialization format used when sending event messages.
...
+
+.. setting:: events_executable
+
+``events_executable``
+~~~~~~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal `python` executable path for :program:`celery events` to use when deaemonizing (default to :data:`sys.executable`).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3235]:
```suggestion
An optional `python` executable path for :program:`celery events` to use when deaemonizing (defaults to :data:`sys.executable`).
```

File: docs/userguide/configuration.rst [hunk 7]
```diff
@@ -3448,6 +3503,62 @@ Default: ``"kombu.asynchronous.hub.timer:Timer"``.
...
 Default is or set by the pool implementation.
 
+.. setting:: worker_logfile
+
+``worker_logfile``
+~~~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal file path for :program:`celery worker` to log into (default to `stdout`).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3513]:
```suggestion
An optional file path for :program:`celery worker` to log into (defaults to `stdout`).
```

File: docs/userguide/configuration.rst [hunk 8]
```diff
@@ -3448,6 +3503,62 @@ Default: ``"kombu.asynchronous.hub.timer:Timer"``.
...
+
+.. setting:: worker_pidfile
+
+``worker_pidfile``
+~~~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal file path for :program:`celery worker` to create/store it PID file (default to no PID file created).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3522]:
```suggestion
An optional file path for :program:`celery worker` to create/store its PID file (defaults to no PID file created).
```

File: docs/userguide/configuration.rst [hunk 9]
```diff
@@ -3448,6 +3503,62 @@ Default: ``"kombu.asynchronous.hub.timer:Timer"``.
...
+
+.. setting:: worker_uid
+
+``worker_uid``
+~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal user ID to use when :program:`celery worker` daemon drops its privileges (default to no UID change).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3531]:
```suggestion
An optional user ID to use when :program:`celery worker` daemon drops its privileges (defaults to no UID change).
```

File: docs/userguide/configuration.rst [hunk 10]
```diff
@@ -3448,6 +3503,62 @@ Default: ``"kombu.asynchronous.hub.timer:Timer"``.
...
+
+.. setting:: worker_gid
+
+``worker_gid``
+~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal group ID to use when :program:`celery worker` daemon drops its privileges (default to no GID change).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3540]:
```suggestion
An optional group ID to use when :program:`celery worker` daemon drops its privileges (defaults to no GID change).
```

File: docs/userguide/configuration.rst [hunk 11]
```diff
@@ -3448,6 +3503,62 @@ Default: ``"kombu.asynchronous.hub.timer:Timer"``.
...
+
+.. setting:: worker_umask
+
+``worker_umask``
+~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal `umask` to use when :program:`celery worker` creates files (log, pid...) when daemonizing.
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3549]:
```suggestion
An optional `umask` to use when :program:`celery worker` creates files (log, pid...) when daemonizing.
```

File: docs/userguide/configuration.rst [hunk 12]
```diff
@@ -3448,6 +3503,62 @@ Default: ``"kombu.asynchronous.hub.timer:Timer"``.
...
+
+.. setting:: worker_executable
+
+``worker_executable``
+~~~~~~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal `python` executable path for :program:`celery worker` to use when deaemonizing (default to :data:`sys.executable`).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3558]:
```suggestion
An optional `python` executable path for :program:`celery worker` to use when deaemonizing (defaults to :data:`sys.executable`).
```

File: docs/userguide/configuration.rst [hunk 13]
```diff
@@ -3534,3 +3645,57 @@ Default: None.
...
 are past due will always run immediately.
+
+.. setting:: beat_logfile
+
+``beat_logfile``
+~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal file path for :program:`celery beat` to log into (default to `stdout`).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3656]:
```suggestion
An optional file path for :program:`celery beat` to log into (defaults to `stdout`).
```

File: docs/userguide/configuration.rst [hunk 14]
```diff
@@ -3534,3 +3645,57 @@ Default: None.
...
+
+.. setting:: beat_pidfile
+
+``beat_pidfile``
+~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal file path for :program:`celery beat` to create/store it PID file (default to no PID file created).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3665]:
```suggestion
An optional file path for :program:`celery beat` to create/store it PID file (defaults to no PID file created).
```

File: docs/userguide/configuration.rst [hunk 15]
```diff
@@ -3534,3 +3645,57 @@ Default: None.
...
+
+.. setting:: beat_uid
+
+``beat_uid``
+~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal user ID to use when beat :program:`celery beat` drops its privileges (default to no UID change).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3674]:
```suggestion
An optional user ID to use when beat :program:`celery beat` drops its privileges (defaults to no UID change).
```

File: docs/userguide/configuration.rst [hunk 16]
```diff
@@ -3534,3 +3645,57 @@ Default: None.
...
+
+.. setting:: beat_gid
+
+``beat_gid``
+~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal group ID to use when :program:`celery beat` daemon drops its privileges (default to no GID change).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3683]:
```suggestion
An optional group ID to use when :program:`celery beat` daemon drops its privileges (defaults to no GID change).
```

File: docs/userguide/configuration.rst [hunk 17]
```diff
@@ -3534,3 +3645,57 @@ Default: None.
...
+
+.. setting:: beat_umask
+
+``beat_umask``
+~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal `umask` to use when :program:`celery beat` creates files (log, pid...) when daemonizing.
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3692]:
```suggestion
An optional `umask` to use when :program:`celery beat` creates files (log, pid...) when daemonizing.
```

File: docs/userguide/configuration.rst [hunk 18]
```diff
@@ -3534,3 +3645,57 @@ Default: None.
...
+
+.. setting:: beat_executable
+
+``beat_executable``
+~~~~~~~~~~~~~~~~~~~
+
+Default: :const:`None`
+
+An optionnal `python` executable path for :program:`celery beat` to use when deaemonizing (default to :data:`sys.executable`).
```
Comment (thedrow) [inline docs/userguide/configuration.rst:3701]:
```suggestion
An optional `python` executable path for :program:`celery beat` to use when deaemonizing (defaults to :data:`sys.executable`).
```

**Review by thedrow — CHANGES_REQUESTED** [review]
There are typos in the documentation. Please review and commit the changes.

**Review by Nusnus — APPROVED** [review]
LGTM

**Comment by Nusnus** [comment]
> In addition, all those new configuration settings should be marked with `.. versionadded:: 5.4`

@noirbizarre thank you for this PR!
Please add the missing 5.4 doc, and I'll merge it right away.

**Comment by noirbizarre** [comment]
Hi 👋🏼 
Sorry for the lack of response lately (overbooked, holidays, children... life). I'll read the entire thread and update the PR accordingly by the end of week (maybe tonight if the remaining tasks are quick) !

**Comment by Nusnus** [comment]
> Hi 👋🏼 Sorry for the lack of response lately (overbooked, holidays, children... life). I'll read the entire thread and update the PR accordingly by the end of week (maybe tonight if the remaining tasks are quick) !

Amazing!
Tag me when you’re done and I’ll finish the job on our side.

Thank you!

**Comment by noirbizarre** [comment]
There it is ! Version added markers for 5.4 added on all settings 👍🏼

Thanks for the review and the merge 🙏🏼

**Comment by Nusnus** [comment]
> There it is ! Version added markers for 5.4 added on all settings 👍🏼
> 
> Thanks for the review and the merge 🙏🏼 

Thank you!!
Unfortunately there's one last detail my friend, the version needs to be first under each title and not after the documentation text.

It's important to keep consistency (see other places) so I hope it's not too annoying to ask for this last fix please @noirbizarre 🙏

**Review by auvipy — APPROVED** [review]
looks good. just the remaining review issues need to be addressed

**Comment by Nusnus** [comment]
> looks good. just the remaining review issues need to be addressed

Fixed my own comment here: #8802

@noirbizarre all good ❤️

**Comment by Nusnus** [comment]
😃
```bash
make -C docs changes
open docs/_build/changes/index.html
```
CleanShot 2024-01-18 at 01 37 34@2x

**Pull request merged**